### PR TITLE
Use conda-pack to simplify Dockerfile

### LINKED ISSUE #215: Consider using conda-pack to reduce unnecessary rebuilds of large build layers

### PR DESCRIPTION
**Pull request by wphicks**
Use conda-pack to move conda environments from dedicated conda installation layers to condaless layers where they will be used. Reduce unnecessary cache invalidation for conda installation layers. Resolve #215 .

### CODE REVIEW

File: ci/gitlab/build.sh [hunk 1]
```diff
@@ -110,7 +110,7 @@ docker run \
   -v "${MODEL_DIR}:/qa/L0_e2e/model_repository" \
   -v "${CPU_MODEL_DIR}:/qa/L0_e2e/cpu_model_repository" \
   $MODEL_BUILDER_IMAGE \
-  bash -c 'conda run -n triton_test /qa/generate_example_models.sh'
+  bash -c 'source /conda/test/bin/activate && /qa/generate_example_models.sh'
```
[the remark this answers is not in the document]
Comment (wphicks) [reply ci/gitlab/build.sh:113]: Which bit? `conda-pack` itself? Latest commit was 10 days ago, so I'm assuming not?

[the remark this answers is not in the document]
Comment (divyegala) [reply ci/gitlab/build.sh:113]: Activating a conda environment with `source`. But it seems like we are running the script of `activate` itself, which should be okay. I'll defer the actual review to someone with more experience for this PR

[the remark this answers is not in the document]
Comment (wphicks) [reply ci/gitlab/build.sh:113]: Ah, I understand what you're saying! That's not the usual `conda` activation but the activation script for the `conda-pack`ed environment. That's the usual method for activating such environments.

[the remark this answers is not in the document]
Comment (dantegd) [reply ci/gitlab/build.sh:113]: Actually I think this is just a matter of `conda-pack` using the older deprecated style of activating conda environments with source activate, just tried it locally: 

```bash
(pytorch-hacking) ➜  pytorch conda-pack
Collecting packages...
Packing environment at '/Users/galahad/mambaforge/envs/pytorch-hacking' to 'pytorch-hacking.tar.gz'
[########################################] | 100% Completed | 15.4s
(base) ➜  pytorch tar -xf pytorch-hacking.tar.gz -C env
(base) ➜  pytorch conda activate env/
(/Users/galahad/Projects/pytorch/env) ➜  pytorch python
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:24:38)
[Clang 12.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>>
```

I would probably lean on using `conda activate`

File: ops/Dockerfile [hunk 2]
```diff
@@ -15,38 +15,62 @@ ARG USE_CLIENT_WHEEL=0
...
-    && rm -f Miniconda3-latest-Linux-$(uname -m).sh 
-
+FROM condaforge/mambaforge as conda-base
 COPY ./ops/gpuci_conda_retry /usr/bin/gpuci_conda_retry
 COPY ./ops/gpuci_mamba_retry /usr/bin/gpuci_mamba_retry
 RUN chmod +x /usr/bin/gpuci_conda_retry /usr/bin/gpuci_mamba_retry
 
-RUN gpuci_conda_retry install -c conda-forge mamba>=0.22
+RUN mkdir /conda
+RUN gpuci_mamba_retry install -c conda-forge conda-pack
```
Comment (dantegd) [inline ops/Dockerfile:24]: I think pinning conda-pack here (0.7?) could save us some head aches

**Review by dantegd — CHANGES_REQUESTED** [review]
Things look good, just the question of using `conda activate` and pinning the conda-pack version

File: ci/gitlab/build.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply ci/gitlab/build.sh:113]: The problem is that one of the things we're trying to do is eliminate `conda` as a dependency in the steps where we make use of those environments. Would we be okay with leaving this as is until `conda-pack` introduces a new recommended workflow for situations where `conda` is not available?

**Review by dantegd — APPROVED** [review]

**Pull request merged**